cmd/cmd_test.go

```diff
@@ -42,13 +74,315 @@ func (s *cmdSuite) TestDistroSupportsReExec(c *C) {
+
+	c.Check(cmd.InternalToolPath("potato"), Equals, p)
+}
+
+func (s *cmdSuite) TestInternalToolPathInOldCore(c *C) {
+	d := c.MkDir()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(d, c.MkDir())()
+	defer cmd.MockVersion("2")()
+
+	p := filepath.Join(d, "/usr/lib/snapd/potato")
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(filepath.Join(d, "/usr/lib/snapd/info"), []byte("VERSION=9999"), 0644), IsNil)
+
+	c.Check(cmd.InternalToolPath("potato"), Equals, p)
+}
+
+func (s *cmdSuite) TestInternalToolPathInBothPrefersNore(c *C) {
+	dNew := c.MkDir()
+	dOld := c.MkDir()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(dOld, dNew)()
+	defer cmd.MockVersion("2")()
+
+	pNew := filepath.Join(dNew, "/usr/lib/snapd/potato")
+	c.Assert(os.MkdirAll(pNew, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(filepath.Join(dNew, "/usr/lib/snapd/info"), []byte("VERSION=9999"), 0644), IsNil)
+
+	pOld := filepath.Join(dOld, "/usr/lib/snapd/potato")
+	c.Assert(os.MkdirAll(pOld, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(filepath.Join(dOld, "/usr/lib/snapd/info"), []byte("VERSION=9999"), 0644), IsNil)
+
+	c.Check(cmd.InternalToolPath("potato"), Equals, pNew)
+}
+
+func (s *cmdSuite) TestInternalToolPathDisabledByEnv(c *C) {
+	os.Setenv("SNAP_REEXEC", "0")
+	defer os.Unsetenv("SNAP_REEXEC")
+
+	c.Check(cmd.InternalToolPath("potato"), Equals, filepath.Join(dirs.DistroLibExecDir, "potato"))
+}
+
+func (s *cmdSuite) TestInternalToolPathDisabledByDistro(c *C) {
+	defer release.MockOnClassic(false)()
+
+	c.Check(cmd.InternalToolPath("potato"), Equals, filepath.Join(dirs.DistroLibExecDir, "potato"))
+}
+
+func (s *cmdSuite) TestInternalToolPathInNewCoreDisabledByNoReExecSupportInCore(c *C) {
+	d := c.MkDir()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(c.MkDir(), d)()
+
+	p := filepath.Join(d, "/usr/lib/snapd/potato")
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(filepath.Join(d, "/usr/lib/snapd/info"), []byte("VERSION=0"), 0644), IsNil)
+
+	c.Check(cmd.InternalToolPath("potato"), Equals, filepath.Join(dirs.DistroLibExecDir, "potato"))
+}
+
+func (s *cmdSuite) TestExecInCoreSnap(c *C) {
+	d := c.MkDir()
+	selfExe := filepath.Join(d, "self-exe")
+	defer cmd.MockVersion("2")()
+	defer cmd.MockSelfExe(selfExe)()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(c.MkDir(), d)()
+	c.Assert(os.Symlink("/bin/true", selfExe), IsNil)
+	c.Assert(os.MkdirAll(filepath.Join(d, "/bin/true"), 0755), IsNil)
+	p := d + "/usr/lib/snapd"
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(p+"/info", []byte("VERSION=9999"), 0644), IsNil)
+
+	c.Check(cmd.ExecInCoreSnap, PanicMatches, `>exec of "[^"]+/bin/true" in tests<`)
```

Comment:
Just curious, why panic here? I.e. what does it give us over just running the subsequent checks (s.execCalled, s.lastExecArgv0 etc)?